sftp/sftp_test.go

```diff
@@ -0,0 +1,284 @@
+package sftp
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/rudderlabs/rudder-go-kit/sftp/mock_sftp"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/sshserver"
+)
+
+func TestSSHClientConfig(t *testing.T) {
+	// Read private key
+	privateKey, err := os.ReadFile("testdata/ssh/test_key")
+	require.NoError(t, err)
+
+	type testCase struct {
+		description   string
+		config        *SSHConfig
+		expectedError error
+	}
+
+	testCases := []testCase{
+		{
+			description:   "WithNilConfig",
+			config:        nil,
+			expectedError: fmt.Errorf("config should not be nil"),
+		},
+		{
+			description: "WithEmptyHostName",
+			config: &SSHConfig{
+				HostName:   "",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: fmt.Errorf("hostname should not be empty"),
+		},
+		{
+			description: "WithEmptyPort",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: fmt.Errorf("port should not be empty"),
+		},
+		{
+			description: "WithPassword",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: nil,
+		},
+		{
+			description: "WithPrivateKey",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "keyAuth",
+				PrivateKey: string(privateKey),
+			},
+			expectedError: nil,
+		},
+		{
+			description: "WithUnsupportedAuthMethod",
+			config: &SSHConfig{
+				HostName:   "HostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "invalidAuth",
+				PrivateKey: "somePrivateKey",
+			},
+			expectedError: fmt.Errorf("unsupported authentication method"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			sshConfig, err := sshClientConfig(tc.config)
+			if tc.expectedError != nil {
+				assert.EqualError(t, tc.expectedError, err.Error())
+				assert.Nil(t, sshConfig)
+			} else {
+				assert.Nil(t, err)
+				assert.NotNil(t, sshConfig)
+			}
+		})
+	}
+}
+
+func TestUpload(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	// Create local and remote directories within the temporary directory
+	baseDir := t.TempDir()
+	localDir := filepath.Join(baseDir, "local")
+	remoteDir := filepath.Join(baseDir, "remote")
+	err := os.MkdirAll(localDir, 0o755)
+	require.NoError(t, err)
+	err = os.MkdirAll(remoteDir, 0o755)
+	require.NoError(t, err)
```

Comment:
Have a look at this function, you might like it :slightly_smiling_face: 

```suggestion
	localDir, err := os.MkdirTemp("", t.Name())
	require.NoError(t, err)
	remoteDir, err := os.MkdirTemp("", t.Name())
	require.NoError(t, err)
```

I tried it myself now and it created them correctly:

```go
t.Log(localDir) // /tmp/TestUpload2172697112 
t.Log(remoteDir) // /tmp/TestUpload1905411585

Reply:
There is also one to create a temporary file: https://pkg.go.dev/os#CreateTemp

Reply:
- Used `MkdirTemp`
- I think `CreateTemp` is not suitable for our testcases. As it appends the **random** string to the given fileName and returns the file pointer. We need the **localFilePath** and **remoteFilePath** for comparing the file content.